mat64/io.go

```diff
@@ -5,14 +5,159 @@
+//
+// Limited checks on the validity of the binary input are performed:
+//  - matrix.ErrShape is returned if the number of rows or columns is negative,
+//  - matrix.ErrShape is returned if the resulting Dense matrix is too
+//  big for the current architecture (e.g. a 16GB matrix written by a
+//  64b application and read back from a 32b application.)
+func (m *Dense) UnmarshalBinary(data []byte) error {
+	if !m.isZero() {
+		panic("mat64: unmarshal into non-zero matrix")
+	}
+
+	buf := bytes.NewReader(data)
+	var rows int64
+	err := binary.Read(buf, binary.LittleEndian, &rows)
+	if err != nil {
+		return err
+	}
+	var cols int64
+	err = binary.Read(buf, binary.LittleEndian, &cols)
+	if err != nil {
+		return err
+	}
+
+	m.mat.Rows = int(rows)
+	m.mat.Cols = int(cols)
+	m.mat.Stride = int(cols)
+	m.capRows = int(rows)
+	m.capCols = int(cols)
+
+	// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+	const maxLen = int64(int(^uint(0) >> 1))
+
+	size := rows * cols
+	if rows < 0 || cols < 0 || int(size) < 0 || size > maxLen {
+		return matrix.ErrShape
+	}
+
+	m.mat.Data = use(m.mat.Data, int(size))
+
+	err = binary.Read(buf, binary.LittleEndian, &m.mat.Data)
+	if err != nil {
+		return err
+	}
+
+	return err
+}
+
+// MarshalBinary encodes the receiver into a binary form and returns the result.
+//
+// Vector is little-endian encoded as follows:
+//   [0 -  8)  number of elements     (int64)
+//   [8 - ..)  vector's data elements (float64)
+func (v Vector) MarshalBinary() ([]byte, error) {
+	buf := bytes.NewBuffer(make([]byte, 0, sizeInt64+v.n*sizeFloat64))
+	err := binary.Write(buf, binary.LittleEndian, int64(v.n))
+	if err != nil {
+		return nil, err
+	}
+
+	for i := 0; i < v.n; i++ {
+		err = binary.Write(buf, binary.LittleEndian, v.mat.Data[i*v.mat.Inc])
+		if err != nil {
+			return nil, err
+		}
+	}
+	return buf.Bytes(), err
+}
+
+// UnmarshalBinary decodes the binary form into the receiver.
+// It panics if the receiver is a non-zero Vector.
+//
+// See MarshalBinary for the on-disk layout.
+//
+// Limited checks on the validity of the binary input are performed:
+//  - matrix.ErrShape is returned if the number of rows is negative,
+//  - matrix.ErrShape is returned if the resulting Vector is too
+//  big for the current architecture (e.g. a 16GB vector written by a
+//  64b application and read back from a 32b application.)
+func (v *Vector) UnmarshalBinary(data []byte) error {
+	if !v.isZero() {
+		panic("mat64: unmarshal into non-zero vector")
+	}
+
+	buf := bytes.NewReader(data)
+	var n int64
+	err := binary.Read(buf, binary.LittleEndian, &n)
+	if err != nil {
+		return err
+	}
+
+	// maxLen is the biggest slice/array len one can create on a 32/64b platform.
+	const maxLen = int64(int(^uint(0) >> 1))
+
+	if n < 0 || n > maxLen {
+		return matrix.ErrShape
+	}
+
+	v.n = int(n)
+	v.mat.Inc = 1
+	v.mat.Data = use(v.mat.Data, v.n)
+
+	err = binary.Read(buf, binary.LittleEndian, &v.mat.Data)
```

Comment:
turns out it isn't:
 https://golang.org/src/encoding/binary/binary.go#L206

done.